submissions/antonkottans/tjsw-post-oop/index.js

```diff
@@ -0,0 +1,163 @@
+class Creature {
+    constructor({ name, gender, friends = [] }) {
+        this.name = name;
+        this.gender = gender;
+        this.friends = friends;
+        this.keys = ["name", "gender", "friends"];
+    }
+
+    returnFullInfo() {
+        return this.keys
+            .map((key) => {
+                if (this[key]) {
+                    if (key === "friends")
+                        if (this["friends"].length != 0)
+                            return (
+                                "friends:" +
+                                this["friends"]
+                                    .map((friend) => friend.name)
+                                    .join(", ")
+                            );
+                        else return null;
+                    else return key + ":" + this[key];
+                } else return null;
+            })
+            .filter((value) => value)
+            .join(",  ");
+    }
+
+    addFriends(...friends) {
+        if (friends)
+            for (const friend of friends) {
+                this.friends = Array.from(friends);
+            }
+    }
+}
+
+class Animal extends Creature {
+    constructor({ name, gender, friends, paws, tail }) {
+        super({ name, gender, friends });
+        this.paws = paws;
+        //this.species = "animal";
+        if (tail) this.tail = tail;
+        this.keys = [...this.keys, /* "species", */ "paws", "tail"];
+    }
+}
+
+class Human extends Creature {
+    constructor({ name, gender, friends, legs, hands, saying }) {
+        super({ name, gender, friends });
+        this.legs = legs;
+        this.hands = hands;
+        this.species = "human";
```

Comment:
Species is a property shared by all creatures. Hence it should be defined and handled in the base class.
This is one of the points of OOP and also makes code DRY.
Generalized, shared props should be defined in a common base class.